dexios/src/domain/encrypt.rs

```diff
@@ -0,0 +1,348 @@
+                        16, 183, 233, 128, 23, 223, 81, 30, 214, 132, 32, 104, 51, 119, 13, 182,
+                        233, 117, 195, 199, 31, 7, 58, 206, 16, 183, 233, 128, 23, 223, 81, 30,
+                        214, 132, 32, 104, 51, 119, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                        0, 0, 0, 250, 200, 57, 180, 83, 36, 137, 168, 42, 45, 18, 74, 46, 70, 96,
+                        86, 65, 27, 121, 4, 55, 33, 220, 149, 241, 61, 180
+                    ]
+                );
+            }
+            Err(e) => {
+                println!("{:?}", e);
+                unreachable!()
+            }
+        }
+    }
+
+    #[test]
+    fn should_save_header_separately() {
+        let mut input_content = b"Hello world";
+        let input_cur = Cursor::new(&mut input_content);
+
+        let mut output_content = vec![];
+        let output_cur = Cursor::new(&mut output_content);
+
+        let mut output_header = vec![];
+        let output_header_cur = Cursor::new(&mut output_header);
+
+        let req = Request {
+            reader: RefCell::new(input_cur),
+            writer: RefCell::new(output_cur),
+            header_writer: Some(RefCell::new(output_header_cur)),
+            raw_key: Protected::new(PASSWORD.as_bytes().to_vec()),
+            header_type: HeaderType {
+                version: HeaderVersion::V5,
+                algorithm: Algorithm::XChaCha20Poly1305,
+                mode: Mode::StreamMode,
+            },
+            hashing_algorithm: HashingAlgorithm::Blake3Balloon(5),
+        };
+
+        match execute(req) {
+            Ok(_) => {
+                assert_eq!(
+                    output_content,
+                    vec![
+                        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
```

Comment:
@brxken128 Why do we have to fill content with zeros when we detach the header?

Reply:
@brxken128 I don't understand)

Reply:
we just skip header_size (and fill content with zeros) when we detach the headers